src/crewai/agents/executor.py

```diff
@@ -185,6 +198,55 @@ def _iter_next_step(
             yield AgentStep(action=output, observation=observation)
             return
 
+        except Exception as e:
+            if LLMContextLengthExceededException(str(e))._is_context_limit_error(
+                str(e)
+            ):
+                if "context_length_exceeded" in str(e):
+                    self._logger.log(
+                        "debug",
+                        "Context length exceeded. Asking user if they want to use summarize prompt to fit, this will reduce context length.",
+                        color="yellow",
+                    )
+                    user_choice = click.confirm(
+                        "Context length exceeded. Do you want to summarize the text to fit models context window?"
+                    )
+                    if user_choice:
+                        self._logger.log(
+                            "debug",
+                            "Context length exceeded. Using summarize prompt to fit, this will reduce context length.",
+                            color="bold_blue",
+                        )
+                        intermediate_steps = self._handle_context_length(
+                            intermediate_steps
+                        )
+
+                        output = self.agent.plan(
+                            intermediate_steps,
+                            callbacks=run_manager.get_child() if run_manager else None,
+                            **inputs,
+                        )
+
+                        if isinstance(output, AgentFinish):
+                            yield output
+                        else:
+                            yield AgentStep(action=output, observation=None)
+                        return
+                    else:
+                        self._logger.log(
+                            "debug",
+                            "Context length exceeded. Consider using smaller text or RAG tools from crewai_tools.",
+                            color="red",
+                        )
+                        raise SystemExit(
+                            "Context length exceeded and user opted not to summarize. Consider using smaller text or RAG tools from crewai_tools."
+                        )
+            yield AgentStep(
+                action=AgentAction("_Exception", str(e), str(e)),
+                observation=str(e),
+            )
+            return
```

Comment:
_Suggestion_: I think this whole portion could be a different function. I saw that we already have the _handle_context_length but we can add a _handle_context_length_error and just call inside of this part of the code, because this function is getting bigger than expected.